### PR TITLE
Exec filter during Scan

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/src/arrow.rs [hunk 1]
```diff
@@ -336,6 +336,9 @@ pub trait RecordBatchExt {
 
     /// Get (potentially nested) column by qualified name.
     fn column_by_qualified_name(&self, name: &str) -> Option<&ArrayRef>;
+
+    /// Project the schema over the [RecordBatch].
+    fn project_by_schema(&self, schema: &Schema) -> Result<RecordBatch>;
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/arrow.rs:341]: `RecordBatch` already has a `project` method that takes first level `column id`, which does not support nested projection.

File: rust/src/dataset/scanner.rs [hunk 2]
```diff
@@ -64,17 +66,15 @@ pub struct Scanner {
 
     projections: Schema,
 
-    /// Optional filters.
-    filter: Option<Expr>,
+    /// Optional filters string.
+    filter: Option<String>,
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/dataset/scanner.rs:70]: Yes, call `pyarrow.compute.Expr.to_str()` and pass to rust. User can directly write string filter as well in python.

File: rust/src/datafusion/physical_expr.rs [hunk 3]
```diff
@@ -103,6 +103,31 @@ pub fn col(name: &str) -> Arc<dyn PhysicalExpr> {
     Arc::new(Column::new(name.to_string()))
 }
 
+struct ColumnVisitor {
+    columns: Vec<String>,
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/datafusion/physical_expr.rs:107]: yes

File: rust/src/dataset/scanner.rs [hunk 4]
```diff
@@ -266,35 +264,39 @@ impl Scanner {
...
+                let knn_node = if q.refine_factor.is_some() {
+                    self.flat_knn(knn_node, &q)
+                } else {
+                    knn_node
+                };
 
-                if q.refine_factor.is_some() {
-                    self.knn(take_node, &q)
+                if let Some(filter_expression) = filter_expr {
+                    self.filter_node(filter_expression, knn_node)?
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/dataset/scanner.rs:275]: For knn, the limit is already `top-k * num_refines`,

File: rust/src/dataset/scanner.rs [hunk 5]
```diff
@@ -364,6 +356,19 @@ impl Scanner {
...
+
+    fn filter_node(
+        &self,
+        filter: Arc<dyn PhysicalExpr>,
+        plan: Arc<dyn ExecutionPlan>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        let filter_node = Arc::new(FilterExec::try_new(filter, plan)?);
+        Ok(Arc::new(LocalTakeExec::new(
+            filter_node,
+            self.dataset.clone(),
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/dataset/scanner.rs:368]: yes

File: rust/src/io/exec/scan.rs [hunk 6]
```diff
@@ -53,40 +48,36 @@ impl LanceStream {
...
         with_row_id: bool,
-    ) -> Self {
+    ) -> Result<Self> {
         let (tx, rx) = mpsc::channel(prefetch_size);
 
         let project_schema = projection.clone();
+        let data_dir = dataset.data_dir();
         let io_thread = tokio::spawn(async move {
-            'outer: for frag in fragments.as_ref() {
+            'outer: for frag in dataset.fragments().as_ref() {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/io/exec/scan.rs:70]: at that time, user will trigger a new query on the "newly created dataset"? so the dataset instance will have the newly append fragment?

File: rust/src/io/exec/take.rs [hunk 7]
```diff
@@ -178,3 +181,169 @@ impl ExecutionPlan for GlobalTakeExec {
         todo!()
     }
 }
+
+pub struct LocalTake {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/io/exec/take.rs:185]: Oh, no it is shared within the dataset. The execute plan graph lasts for one query entirely.

**Review by changhiskhan — APPROVED** [review]

**Pull request merged**